src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -517,89 +477,117 @@ internal unsafe ValueTask ConnectAsync(CancellationToken cancellationToken = def
+                        targetHost = _state.TargetHost!;
+                        if (IPAddress.TryParse(dnsHost, out IPAddress? address))
+                        {
+                            // This is form of IPAddress and _state.TargetHost is set to different string
+                            Debug.Assert(!Monitor.IsEntered(_state), "!Monitor.IsEntered(_state)");
+                            MsQuicParameterHelpers.SetIPEndPointParam(MsQuicApi.Api, _state.Handle, QUIC_PARAM_CONN_REMOTE_ADDRESS, new IPEndPoint(address, port));
+                        }
+                        else
+                        {
+                            IPAddress[] addresses = Dns.GetHostAddressesAsync(dnsHost, cancellationToken).GetAwaiter().GetResult();
+                            cancellationToken.ThrowIfCancellationRequested();
+                            if (addresses.Length == 0)
+                            {
+                                throw new SocketException((int)SocketError.HostNotFound);
+                            }
+                            Debug.Assert(!Monitor.IsEntered(_state), "!Monitor.IsEntered(_state)");
+                            // We can do something better than just using first IP but that is what
+                            // MsQuic does today anyway.
+                            MsQuicParameterHelpers.SetIPEndPointParam(MsQuicApi.Api, _state.Handle, QUIC_PARAM_CONN_REMOTE_ADDRESS, new IPEndPoint(addresses[0], port));
+                        }
                     }
                     else
                     {
-                        IPAddress[] addresses = Dns.GetHostAddressesAsync(dnsHost, cancellationToken).GetAwaiter().GetResult();
-                        cancellationToken.ThrowIfCancellationRequested();
-                        if (addresses.Length == 0)
-                        {
-                            throw new SocketException((int)SocketError.HostNotFound);
-                        }
-                        Debug.Assert(!Monitor.IsEntered(_state), "!Monitor.IsEntered(_state)");
-                        // We can do something better than just using first IP but that is what
-                        // MsQuic does today anyway.
-                        MsQuicParameterHelpers.SetIPEndPointParam(MsQuicApi.Api, _state.Handle, QUIC_PARAM_CONN_REMOTE_ADDRESS, new IPEndPoint(addresses[0], port));
+                        // We defer everything to MsQuic.
+                        targetHost = dnsHost;
                     }
                 }
                 else
                 {
-                    // We defer everything to MsQuic.
-                    targetHost = dnsHost;
+                    throw new ArgumentException($"Unsupported remote endpoint type '{_remoteEndPoint.GetType()}'.");
+                }
+
+                IntPtr pTargetHost = Marshal.StringToCoTaskMemAnsi(targetHost);
```

Comment:
Ansi is the right thing here, even on Linux?

Reply:
We want ASCII here - https://www.rfc-editor.org/rfc/rfc6066.html#section-3:
> The hostname is represented as a byte
   string using ASCII encoding without a trailing dot.  This allows the
   support of internationalized domain names through the use of A-labels
   defined in [[RFC5890](https://www.rfc-editor.org/rfc/rfc5890)].


Is there something better to use? I'm not very well versed in this space, so please feel free to suggest better way to do this.

Note that we don not support internationalized names yet.